### ISSUE TITLE
NullPointerException in Component.newComponentConfiguration when multiple component instances are bound to the same service and have the same target property

### ISSUE THREAD

**Issue by magicmike2248** [knopflerfish/knopflerfish.org#17]
I have two components defined: **TargetService** and **Component**.

## TargetService ##

<pre>
public interface TargetService {
}

@org.osgi.service.component.annotations.Component(
        name = "TargetService",

        configurationPolicy = ConfigurationPolicy.REQUIRE,
        immediate = true)
public class TargetServiceImpl implements TargetService {

    @Activate
    public void activate(Map properties) {
        println("=== TargetService Activate " + properties.get("number") + " ===");
    }

    @Deactivate
    public void deactivate(Map properties) {
        println("=== TargetService Deactivate " + properties.get("number") + " ===");
    }

    private void println(String message) {
        System.out.println(message);
    }
}
</pre>

## Component ##

<pre>
@org.osgi.service.component.annotations.Component(
        name = "Component",
        configurationPolicy = ConfigurationPolicy.REQUIRE,
        immediate = true)
public class Component {

    private TargetService target;

    @Reference(unbind = "unbindTargetService")
    public void bindTargetService(TargetService target) {
        this.target = target;
    }

    public void unbindTargetService(TargetService target) {
        this.target = null;
    }

    @Activate
    public void activate(Map properties) {
        println("=== Component Activate " + properties.get("number") + " ===");
        println(properties);
    }

    @Deactivate
    public void deactivate(Map properties) {
        println("=== Component Deactivate " + properties.get("number") + " ===");
        println(properties);
    }

    private void println(Map<String,String> properties) {
        if (properties != null) {
            Iterator it = properties.entrySet().iterator( );
            while (it.hasNext()) {
                println(it.next().toString());
            }
        }
    }

    private void println(String message) {
        System.out.println(message);
    }
}
</pre>

I want Components 101 and 102 to be bound to TargetService 1 and Component 201 to be bound to TargetService 2, like so:

![chart](https://cloud.githubusercontent.com/assets/11747700/22335658/99a66b60-e3f0-11e6-84f9-bbf182b59a6e.png)

Let's create their configurations:

<pre>
> enter configuration
configuration> create -f Component
configuration> set TargetService.target (number=1)
configuration> set number 101
configuration> save
configuration> create -f Component
configuration> set TargetService.target (number=1)
configuration> set number 102
configuration> save
configuration> create -f Component
configuration> set TargetService.target (number=2)
configuration> set number 201
configuration> save
configuration> show
[-] Component._0
 factory PID: Component
 location: -
 change count: 1
 properties:
  TargetService.target= (number=1)
  number= 101
  service.factoryPid= Component
  service.pid= Component._0
[-] Component._1
 factory PID: Component
 location: -
 change count: 1
 properties:
  TargetService.target= (number=1)
  number= 102
  service.factoryPid= Component
  service.pid= Component._1
[-] Component._2
 factory PID: Component
 location: -
 change count: 1
 properties:
  TargetService.target= (number=2)
  number= 201
  service.factoryPid= Component
  service.pid= Component._2
</pre>

Components 101 and 102 should be activated as soon as their references for TargetService 1 are satisfied.
But when TargetService configuration becomes available, a NullPointerException occurs.

<pre>
configuration> create -f TargetService
configuration> set number 1
configuration> save
configuration> [stderr] ## DEBUG: errors - FrameworkErrorEvent bundle #28
[stderr] ## DEBUG: errors - FrameworkErrorEvent throwable: 
[stderr] java.lang.NullPointerException
[stderr] 	at org.knopflerfish.bundle.component.Component.newComponentConfiguration(Component.java:774)
[stderr] 	at org.knopflerfish.bundle.component.Component.newComponentConfigurations(Component.java:752)
[stderr] 	at org.knopflerfish.bundle.component.Component.satisfied(Component.java:483)
[stderr] 	at org.knopflerfish.bundle.component.Component.refAvailable(Component.java:714)
[stderr] 	at org.knopflerfish.bundle.component.Reference.refAvailable(Reference.java:456)
[stderr] 	at org.knopflerfish.bundle.component.ReferenceListener.serviceChanged(ReferenceListener.java:497)
[stderr] 	at org.knopflerfish.bundle.component.ReferenceListener.serviceEvent(ReferenceListener.java:455)
[stderr] 	at org.knopflerfish.bundle.component.ComponentServiceListener.serviceChanged(ComponentServiceListener.java:71)
[stderr] 	at org.knopflerfish.framework.Listeners.serviceChanged(Listeners.java:430)
[stderr] 	at org.knopflerfish.framework.PermissionOps.callServiceChanged(PermissionOps.java:340)
[stderr] 	at org.knopflerfish.framework.Services.register(Services.java:179)
[stderr] 	at org.knopflerfish.framework.BundleContextImpl.registerService(BundleContextImpl.java:274)
[stderr] 	at org.knopflerfish.bundle.component.ComponentService.registerService(ComponentService.java:72)
[stderr] 	at org.knopflerfish.bundle.component.ComponentConfiguration.registerComponentService(ComponentConfiguration.java:376)
[stderr] 	at org.knopflerfish.bundle.component.ImmediateComponent.activateComponentConfiguration(ImmediateComponent.java:60)
[stderr] 	at org.knopflerfish.bundle.component.Component.satisfied(Component.java:485)
[stderr] 	at org.knopflerfish.bundle.component.Component.resolvedConstraint(Component.java:582)
[stderr] 	at org.knopflerfish.bundle.component.Component.cmConfigUpdated(Component.java:645)
[stderr] 	at org.knopflerfish.bundle.component.CMConfig.cmPidSet(CMConfig.java:241)
[stderr] 	at org.knopflerfish.bundle.component.CMConfig.access$100(CMConfig.java:53)
[stderr] 	at org.knopflerfish.bundle.component.CMConfig$CMPid.set(CMConfig.java:341)
[stderr] 	at org.knopflerfish.bundle.component.CMConfig$CMPid.configUpdated(CMConfig.java:391)
[stderr] 	at org.knopflerfish.bundle.component.CMHandler.configurationEvent(CMHandler.java:116)
[stderr] 	at org.knopflerfish.bundle.cm.ListenerEvent.sendEvent(ListenerEvent.java:84)
[stderr] 	at org.knopflerfish.bundle.cm.ListenerEventQueue.run(ListenerEventQueue.java:90)
[stderr] 	at java.lang.Thread.run(Thread.java:745)
[stdout] === TargetService Activate 1 ===
</pre>

I think the problem is in how Reference.update method works.

<pre>
  void update(String ccid, Map<String, Object> dict, boolean useNoId) {
    boolean before = isSatisfied();
    boolean doCheck = updateMinCardinality(dict) && isSatisfied() != before;
    if (listener != null) {
      // We only have one listener, check if it still is true;
      if (listener.checkTargetChanged(ccid, dict)) {
        if (listener.isOnlyId(useNoId ? Component.NO_CCID : ccid)) {
          // Only one ccid change listener target
          listener.setTarget(ccid, dict);
        } else {
          // We have multiple listener we need multiple listeners
          factoryListeners = new TreeMap<String, ReferenceListener>();
          for (String p : listener.getIds()) {
            factoryListeners.put(p, listener);
          }
          listener = null;
          // NYI, optimize, we don't have to checkTargetChanged again
        }
      } else if (useNoId) {
        // No change, just make sure that ccid is registered
        listener.addId(ccid, true);
      }
    }
    ...
</pre>

The first call to this method (ccid = "Component._0") sets the target to (number=1).
The second call to this method (with ccid = "Component._1") has no effect: 
listener.checkTargetChanged return false, because target hasn't changed since the previous call ((number=1)),
and useNoId is false, so neither branch there is taken and ccid "Component._1" is lost.
The third call creates factoryListeners with only two entries:
<pre>
factoryListeners = {TreeMap@3447}  size = 2
 0 = {TreeMap$Entry@3453} "Component._0" -> "ReferenceListener(Reference TargetService in Immediate component: Component, target=(number=1))"
 1 = {TreeMap$Entry@3454} "Component._2" -> "ReferenceListener(Reference TargetService in Immediate component: Component, target=(number=2))"
</pre>

Later, Component.newComponentConfiguration("Component._1".....) calls ref.getListener("Component._1").numAvailable( ), but ref.getListener("Component._1") returns null.

**Comment by janste63** [comment]
Which version of the component bundle is used?

**Comment by magicmike2248** [comment]
6.0.2, built from develop

**Comment by magicmike2248** [comment]
Also reproducible in component_all-5.0.3.jar

**Comment by janste63** [comment]
Mike,
I think you're right in your analysis, we need to add the new id.
Could you do the following changes and re-test?

```
--- a/osgi/bundles/component/src/org/knopflerfish/bundle/component/Reference.java
+++ b/osgi/bundles/component/src/org/knopflerfish/bundle/component/Reference.java
@@ -314,9 +314,9 @@ class Reference implements org.apache.felix.scr.Reference
           listener = null;
           // NYI, optimize, we don't have to checkTargetChanged again
         }
-      } else if (useNoId) {
-        // No change, just make sure that ccid is registered
-        listener.addId(ccid, true);
+      } else {
+        // No change, add new ccid or replace if was NO_CCID.
+        listener.addId(ccid, useNoId);
       }
     }
     if (factoryListeners != null) {

```

**Comment by magicmike2248** [comment]
I applied the changes, and it started working a little bit better, NPE is not thrown when component instances are created for the first time, but if I then delete their configurations and create them again, NPE is thrown. 
I tested it with a larger setup, ~50 components of different types with complex dependencies between them.

**Comment by janste63** [comment]
Fixed this issue in component version 6.0.3. The new version also contains a number of other fixes related to factory configurations. I'll closes this now. If there are any other remaining problems, create a new issue.
